File: src/main/java/api/routes/requestchannel/RequestChannelController.java
package api.routes.requestchannel;

import api.routes.requestchannel.dto.CreateRequestChannelDto;
import api.routes.requestchannel.dto.ToggleRequestChannelButtonDto;
import api.routes.requestchannel.dto.ToggleRequestChannelButtonsDto;
import api.utils.GeneralResponse;
import api.utils.ReqChannelCreationResponse;
import jakarta.validation.Valid;
import lombok.AllArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/reqchannel")
@AllArgsConstructor
public class RequestChannelController {

    private final RequestChannelService service;

    @PostMapping("")
    public ResponseEntity<ReqChannelCreationResponse> createChannel(@Valid @RequestBody CreateRequestChannelDto createRequestChannelDto) {
        return service.createChannel(createRequestChannelDto);
    }

    @PostMapping("/button")
    public ResponseEntity<GeneralResponse> toggleButtonVisibility(@Valid @RequestBody ToggleRequestChannelButtonDto toggleRequestChannelButtonDto) {
        return service.toggleButton(toggleRequestChannelButtonDto);
    }

    @PostMapping("/buttons")
    public ResponseEntity<GeneralResponse> toggleButtonsVisibility(@Valid @RequestBody ToggleRequestChannelButtonsDto toggleRequestChannelButtonDto) {
        return service.toggleButtons(toggleRequestChannelButtonDto);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<GeneralResponse> deleteChannel(@PathVariable String id) {
        return service.deleteChannel(id);
    }
}

File: src/main/kotlin/api/routes/requestchannel/RequestChannelController.kt
package api.routes.requestchannel

import api.plugins.routeWithJwt
import api.routes.requestchannel.dto.CreateRequestChannelDto
import api.routes.requestchannel.dto.ToggleRequestChannelButtonDto
import api.routes.requestchannel.dto.ToggleRequestChannelButtonsDto
import api.utils.respond
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.routing.*
import net.dv8tion.jda.api.sharding.ShardManager
import org.koin.ktor.ext.inject

fun Routing.requestChannel() {
    val shardManager: ShardManager by inject()
    val service = RequestChannelService(shardManager)

    routeWithJwt("/reqchannel") {
        post {
            val dto = call.receive<CreateRequestChannelDto>()
            call.respond(service.createChannel(dto))
        }

        patch("button") {
            val dto = call.receive<ToggleRequestChannelButtonDto>()
            call.respond(service.toggleButton(dto))
        }

        patch("buttons") {
            val dto = call.receive<ToggleRequestChannelButtonsDto>()
            call.respond(service.toggleButtons(dto))
        }

        delete("{id}") {
            val id = call.parameters["id"]!!
            call.respond(service.deleteChannel(id))
        }
    }
}